cityofaustin/vision-zero
# Remove crashes edits layer

## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/19185
 PR https://github.com/cityofaustin/vision-zero/pull/1551 revealed that there is need for our data model to enable null user overrides to CRIS values. This PR is a proof of concept on the crashes tables that implements the 2 layer data model structure i proposed awhile back in the data model prototyping stage. There are only two tables, the `crashes_cris` table which only contains values straight from cris, and the `crashes` table which acts as the unified table that contains both cris values and user edits if there was a user override. This is the table that users will be updating. We are basically just nixing the whole `_edits` layer.

## Testing

**URL to test:** 
Local

**Steps to test:**
1. Check out this branch, get a fresh replica of prod data. Apply hasura migrations and metadata.
2. Lets test a user edit to the crashes table:
```sql
select rpt_street_name from crashes where id = 1;
select rpt_street_name from crashes_cris where id = 1;
update crashes set rpt_street_name = 'VZ EDIT' where id = 1;
select rpt_street_name from crashes where id = 1;
```
3. Lets test a cris update to that same record, the unified record should show the VZ edit:
```sql
update crashes_cris set rpt_street_name = 'CRIS UPDATE' where id = 1;
select rpt_street_name from crashes where id = 1;
select rpt_street_name from crashes_cris where id = 1;
```
4. Lets test a cris update to a non edited value, the unified record should show the new cris value:
```sql
select rpt_street_name from crashes where id = 2;
select rpt_street_name from crashes_cris where id = 2;
update crashes_cris set rpt_street_name = 'CRIS UPDATE' where id = 2;
select rpt_street_name from crashes where id = 2;
select rpt_street_name from crashes_cris where id = 1;
```
5. Now lets test a user null override. The unified record should show null instead of the cris value
```sql
select rpt_street_name from crashes where id = 10;
select rpt_street_name from crashes_cris where id = 10;
update crashes set rpt_street_name = null where id = 10;
select rpt_street_name from crashes where id = 10;
select rpt_street_name from crashes_cris where id = 10;
```
6. Next we can test the cris import script. You can download some extract zips from the AWS S3 bucket and put them in `/vision-zero/etl/cris_import/extracts` then follow the test steps in  `cris_import/README.md` to run the cris import docker image and run the following command to process those extract zips:
```shell
./cris_import.py --csv
```
7. You can also inspect the `change_log_crashes` and `change_log_crashes_cris` and make sure you are seeing your changes there as expected
8. Test the down migrations


---

#### Ship list
